src/server/frontend_wayland/ext_image_capture_v1.cpp

```diff
@@ -0,0 +1,610 @@
+
+    DamageAmount damage_amount = DamageAmount::full;
+    geom::Rectangle output_space_damage;
+    wayland::Weak<ExtImageCopyCaptureFrameV1> current_frame;
+    std::shared_ptr<scene::SceneChangeNotification> change_notifier;
+};
+
+class ExtImageCopyCaptureFrameV1
+    : public wayland::ImageCopyCaptureFrameV1
+{
+public:
+    ExtImageCopyCaptureFrameV1(wl_resource* resource, ExtImageCopyCaptureSessionV1* session, std::shared_ptr<ExtImageCaptureV1Ctx> const& ctx);
+
+    bool is_ready() const;
+
+    void capture_frame(geom::Rectangle output_space_area, glm::mat2 transform, bool overlay_cursor, geom::Size buffer_size, geom::Rectangle buffer_space_damage);
+
+private:
+    void attach_buffer(wl_resource* buffer) override;
+    void damage_buffer(int32_t x, int32_t y, int32_t width, int32_t height) override;
+    void capture() override;
+
+    void report_result(std::optional<time::Timestamp> captured_time,
+                       geom::Rectangle buffer_space_damage);
+
+    bool capture_has_been_called = false;
+    wayland::Weak<wayland::Buffer> target;
+    geom::Rectangle frame_damage;
+
+    wayland::Weak<ExtImageCopyCaptureSessionV1> const session;
+    std::shared_ptr<ExtImageCaptureV1Ctx> const ctx;
+};
+
+}
+
+auto mf::create_ext_output_image_capture_source_manager_v1(
+    wl_display* display)
+-> std::shared_ptr<wayland::OutputImageCaptureSourceManagerV1::Global>
+{
+    return std::make_shared<ExtOutputImageCaptureSourceManagerV1Global>(display);;
+}
+
+mf::ExtOutputImageCaptureSourceManagerV1Global::ExtOutputImageCaptureSourceManagerV1Global(
+    wl_display* display)
+    : Global{display, Version<1>()}
+{
+}
+
+void mf::ExtOutputImageCaptureSourceManagerV1Global::bind(wl_resource* new_resource)
+{
+    new ExtOutputImageCaptureSourceManagerV1{new_resource};
+}
+
+mf::ExtOutputImageCaptureSourceManagerV1::ExtOutputImageCaptureSourceManagerV1(
+    wl_resource* resource)
+    : wayland::OutputImageCaptureSourceManagerV1{resource, Version<1>()}
+{
+}
+
+void mf::ExtOutputImageCaptureSourceManagerV1::create_source(wl_resource* new_resource, wl_resource* output)
+{
+    auto& output_global = OutputGlobal::from_or_throw(output);
+    new ExtImageCaptureSourceV1{new_resource, output_global, this};
+}
+
+mf::ExtImageCaptureSourceV1::ExtImageCaptureSourceV1(
+    wl_resource* resource, OutputGlobal& output,
+    ExtOutputImageCaptureSourceManagerV1 *manager)
+    : wayland::ImageCaptureSourceV1{resource, Version<1>()},
+      output{&output},
+      manager{manager}
+{
+}
+
+mf::ExtImageCaptureSourceV1* mf::ExtImageCaptureSourceV1::from_or_throw(wl_resource *resource)
+{
+    auto source = dynamic_cast<ExtImageCaptureSourceV1*>(wayland::ImageCaptureSourceV1::from(resource));
+
+    if (!source)
+    {
+        BOOST_THROW_EXCEPTION(std::runtime_error(
+            "ext_image_capture_source_v1@" +
+            std::to_string(wl_resource_get_id(resource)) +
+            " is not a mir::frontend::ExtImageCaptureSourceV1"));
```

Comment:
Non-blocking nit, but `std::format()` probably does this better:
```suggestion
        BOOST_THROW_EXCEPTION(std::runtime_error(
            std::format(
                "ext_image_capture_source_v1@{} is not a mir::frontend::ExtImageCaptureSourceV1",
                wl_resource_get_id(resource)));
```